File: UI/travle_core/lib/src/models/tour_review_response.dart
import 'package:json_annotation/json_annotation.dart';

part 'tour_review_response.g.dart';

/// A rating + comment on a tour, gated to the reviewer's own Completed booking
/// (mirrors the backend `TourReviewResponse`). Tour and user references are
/// flattened to names. Removed reviews only reach admins.
@JsonSerializable()
class TourReviewResponse {
  TourReviewResponse({
    required this.id,
    required this.tourId,
    required this.tourName,
    required this.bookingId,
    required this.userId,
    required this.username,
    required this.authorName,
    required this.rating,
    required this.isRemoved,
    required this.createdAt,
    this.comment,
    this.removedByUserId,
    this.removedByUsername,
    this.removalReason,
    this.modifiedAt,
  });

  final int id;

  final int tourId;
  final String tourName;

  final int bookingId;

  final int userId;
  final String username;
  final String authorName;

  final int rating;
  final String? comment;

  final bool isRemoved;
  final int? removedByUserId;
  final String? removedByUsername;
  final String? removalReason;

  final DateTime createdAt;
  final DateTime? modifiedAt;

  factory TourReviewResponse.fromJson(Map<String, dynamic> json) =>
      _$TourReviewResponseFromJson(json);

  Map<String, dynamic> toJson() => _$TourReviewResponseToJson(this);
}

File: Backend/Travle.Model/Responses/TourReviewResponse.cs
namespace Travle.Model.Responses
{
    /// <summary>
    /// A rating + comment on a tour, gated to the reviewer's own Completed booking. Used by the tour's
    /// reviews list, the organizer's "reviews of my tours" view, and the admin moderation table. Tour and
    /// user references are flattened to names. Removed reviews are excluded from public reads and
    /// aggregates; the moderation view opts them in and reads the removal audit. <see cref="BookingId"/>
    /// stays occupied after removal (no re-review — 03 §3).
    /// </summary>
    public class TourReviewResponse
    {
        public int Id { get; set; }

        public int TourId { get; set; }
        public string TourName { get; set; } = string.Empty;

        public int BookingId { get; set; }

        public int UserId { get; set; }
        public string Username { get; set; } = string.Empty;
        public string AuthorName { get; set; } = string.Empty;

        public int Rating { get; set; }
        public string? Comment { get; set; }

        public bool IsRemoved { get; set; }
        public int? RemovedByUserId { get; set; }
        public string? RemovedByUsername { get; set; }
        public string? RemovalReason { get; set; }

        public DateTime CreatedAt { get; set; }
        public DateTime? ModifiedAt { get; set; }
    }
}
